---
title: "Figure S4.1e,f and S4.2 - reproducibility script"
author: "Wangjie Liu"
date: "2024-03-20"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "~/NAS2/TF-seq/Vincent/TF-seq_proper/")
getwd()
```

```{r}
setwd("~/NAS2/TF-seq/Vincent/TF-seq_proper/")
```

## Libraries & functions

First, I'm loading the required libraries & functions

```{r}
suppressPackageStartupMessages(library(Seurat, lib.loc = "~/.local/share/")) 
suppressPackageStartupMessages(library(data.table))
suppressPackageStartupMessages(library(ggplot2)) 
suppressPackageStartupMessages(library(cowplot)) 
suppressPackageStartupMessages(library(tidyverse)) 
suppressPackageStartupMessages(library(gridExtra)) 
suppressPackageStartupMessages(library(viridis))
suppressPackageStartupMessages(library(RColorBrewer))
suppressPackageStartupMessages(library(monocle3)) # monocle3 version 1.3.4 
# suppressPackageStartupMessages(library(monocle3, lib.loc = "~/.local/share/"))  # 1.0.0
suppressPackageStartupMessages(library(org.Mm.eg.db)) 
suppressPackageStartupMessages(library(clusterProfiler))
suppressPackageStartupMessages(library(pheatmap))
suppressPackageStartupMessages(library(crayon)) 




cat(bold("R"), "version", as.character(version$version.string), "\n")
cat(bold("Seurat"), "version", as.character(packageVersion("Seurat")), "\n")
cat(bold("data.table"), "version", as.character(packageVersion("data.table")), "\n")
cat(bold("ggplot2"), "version", as.character(packageVersion("ggplot2")), "\n")
cat(bold("cowplot"), "version", as.character(packageVersion("cowplot")), "\n")
cat(bold("tidyverse"), "version", as.character(packageVersion("tidyverse")), "\n")
cat(bold("gridExtra"), "version", as.character(packageVersion("gridExtra")), "\n")
cat(bold("viridis"), "version", as.character(packageVersion("viridis")), "\n")
cat(bold("RColorBrewer"), "version", as.character(packageVersion("RColorBrewer")), "\n")
cat(bold("monocle3"), "version", as.character(packageVersion("monocle3")), "\n")
cat(bold("org.Mm.eg.db"), "version", as.character(packageVersion("org.Mm.eg.db")), "\n")
cat(bold("clusterProfiler"), "version", as.character(packageVersion("clusterProfiler")), "\n")
cat(bold("pheatmap"), "version", as.character(packageVersion("pheatmap")), "\n")
# R version R version 4.3.1 (2023-06-16) 
# Seurat version 4.4.0 
# data.table version 1.15.2 
# ggplot2 version 3.5.0 
# cowplot version 1.1.3 
# tidyverse version 2.0.0 
# gridExtra version 2.3 
# viridis version 0.6.5
# RColorBrewer version 1.1.3 
# monocle3 version 1.3.4  
# org.Mm.eg.db version 3.17.0
# clusterProfiler version 4.8.3
# pheatmap version 1.0.12

# monocle3 single-cell pipeline
monocle3_reanalysis <- function(tfoi, data, data.gene_annot, num_dims, res){
  batch.oi <- table(data$batch[data$TF == tfoi]) %>% names()
  cat("batch for ", tfoi, ": ", "\n")
  cat(batch.oi, "\n")
  seu.oi <- subset(data, TF %in% c("D0","D0_confluent", tfoi) & batch %in% batch.oi)
  seu.oi$TF <- factor(seu.oi$TF, levels = c("D0","D0_confluent",tfoi))
  cat("Number of cells for: ", "\n")
  cat(table(seu.oi$TF), "\n")
  
  
  # prepare cds
  cell.meta <- seu.oi@meta.data[, c("nCount_RNA","nFeature_RNA","Phase_corrected","batch","Dose","TF")]
  gene.meta <- data.frame(ens_id = rownames(seu.oi@assays$RNA@counts), gene_short_name = data.gene_annot[rownames(seu.oi@assays$RNA@counts),"gene_symbol"], gene_type = data.gene_annot[rownames(seu.oi@assays$RNA@counts), "biotype"])
  rownames(gene.meta) <- rownames(seu.oi@assays$RNA@counts)
  cds <- new_cell_data_set(expression_data = seu.oi@assays$RNA@counts, 
                         cell_metadata = cell.meta,
                         gene_metadata = gene.meta)
  
  # preprocess
  set.seed(10)
  cds <- preprocess_cds(cds, num_dim = num_dims) # dimensionality of the reduced space, decided according to the variance explained by each PC
  plot_pc_variance_explained(cds) %>% 
    print()
  
  #remove batch
  if (length(batch.oi) > 1){
    cds <- align_cds(cds, alignment_group = "batch")
  }
  #reduce dimension
  cds <- reduce_dimension(cds)
  # clustering
  cds <- cluster_cells(cds, resolution = res) 
  plot_cells(cds, color_cells_by = "TF", show_trajectory_graph = F, cell_size = 2, label_cell_groups = F) %>% 
    print()
  
  # adjust meta data for control and TFoi
  cds$meta.new <- cds@clusters$UMAP$clusters %>% as.character()
  clusters.tfoi <- cds$meta.new[!cds$TF %in% c("D0","D0_confluent")] %>% table()
  clusters.tfoi <- clusters.tfoi[clusters.tfoi >= 10] %>% names() # at least 10 TF cells in that cluster
  cds$meta.new[!cds$meta.new %in% clusters.tfoi] <- "Control"
  
  percent.tfoi <- table(cds$meta.new[cds$TF == tfoi]) / table(cds$meta.new)[names(table(cds$meta.new[cds$TF == tfoi]))]

  
  clusters.tfoi <- intersect(clusters.tfoi, names(percent.tfoi[percent.tfoi > 0.5])) # at least 50% TF cells in that cluster
  cds$meta.new[!cds$meta.new %in% clusters.tfoi] <- "Control"
  cds$meta.new <- factor(cds$meta.new, levels = c("Control", clusters.tfoi))
  plot_cells(cds, color_cells_by = "meta.new", show_trajectory_graph = F, cell_size = 2, label_cell_groups = F) %>% 
    print()

  
  return(cds)
}


# plot TF dose and clusters on UMAP, and TF dose grouped by clusters on violin
plot_dose_clusters <- function(data, tfoi, PATH.UMAP, PATH.violin, cols, output_plot = TRUE){
  # plot dose on UMAP
  p_dose <- plot_cells(data, 
            color_cells_by = "Dose", 
            show_trajectory_graph = F, 
            cell_size = 1, 
            label_cell_groups = F)+
  geom_point(aes(color = cell_color), 
             size = 1, 
             stroke = 1.5)+
  scale_color_continuous(type = "viridis")+
  theme_void()+
  labs(title = paste0(tfoi, " dose"))
  # plot clusters on UMAP
  p_cluster <- plot_cells(data, 
            color_cells_by = "meta.new", 
            show_trajectory_graph = F, 
            cell_size = 1, 
            label_cell_groups = F)+
  geom_point(aes(color = cell_color), 
             size = 1, 
             stroke = 1.5)+
  scale_color_manual(values = cols)+
  theme_void()+
  labs(title = "Cluster")

  
  # plot dose grouped by clusters on violin
  meta.cluster <- colData(data)[, c("meta.new","Dose","TF")] %>% as.data.frame()
  meta.cluster <- meta.cluster[meta.cluster$TF == tfoi & meta.cluster$meta.new != "Control", ]
  p_violin <- ggplot(meta.cluster)+
    geom_violin(mapping = aes(x = meta.new, y = Dose, fill = meta.new))+
    scale_fill_manual(values = c(cols))+
    theme_cowplot()+
    labs(title = tfoi)
  
  # t test or pairwise t test for dose difference in clusters
  clusters.TFoi <- meta.cluster$meta.new %>% as.character() %>% unique()
  num_cluster <- length(clusters.TFoi)
  print(clusters.TFoi)
  if (num_cluster == 2){
    pwc <- t.test(meta.cluster$Dose[meta.cluster$meta.new == clusters.TFoi[1]],meta.cluster$Dose[meta.cluster$meta.new == clusters.TFoi[2]])
  } else if (num_cluster >2){
    pwc <- pairwise.t.test(meta.cluster$Dose, meta.cluster$meta.new, p.adjust.method = "fdr")
    pwc
  }

  cat("t test or pairwise t test:", "\n")
  print(pwc)
  
  # save plots to pdf
  if (output_plot == TRUE){
     ggsave(p_dose + p_cluster, file = PATH.UMAP, width = 10, height = 4)
     ggsave(p_violin, file = PATH.violin, width = 4, height = 4)
  }
  
  p <- list(p_dose, p_cluster, p_violin)
  names(p) <- c("umap.dose","umap.cluster","dose.violin")
  
  return(p)
}

# DE analysis
Calculate_DE <- function(data){
  variable_genes <- top_markers(data, group_cells_by = "meta.new", marker_sig_test = T )
  variable_genes <- variable_genes %>% 
    group_by(cell_group) %>% 
    filter(marker_test_q_value < 0.05) 
  variable_genes$cell_group <- factor(variable_genes$cell_group, levels = levels(data$meta.new))
  return(variable_genes)
}

# exclude overlapped GO terms
exclude_overlapped_GO <- function(percent.overlap = 0.5, df.GOenrich.res){
  # order GO terms by p.adjust
  # so that the intersection check is performed and 
  # GO terms that are both overlapped with others and have lower p values than overlapped terms 
  # will be excluded
  df.GOenrich.res <- df.GOenrich.res[order(df.GOenrich.res$p.adjust, decreasing = T),] 
  go.terms <- df.GOenrich.res$geneID
  go.terms <- str_split(go.terms, "/")
  names(go.terms) <- df.GOenrich.res$ID
  
  terms.overlap <- c()
  nTerms <- length(go.terms)
  if(nTerms > 1){
    for (i in 1:(nTerms-1)){
      Fractions <- lapply(go.terms[i:nTerms], function(x){
        overlap.fract <- length(intersect(x, go.terms[[i]]))/length(x) 
        return(overlap.fract)
      })
      names(Fractions) <- names(go.terms)[i:nTerms]
      Fractions <- unlist(Fractions)
      # if the term has more than 50% overlap with other terms
      if (table(Fractions > percent.overlap)["TRUE"] == 1){ #only itself
        terms.overlap[i] <- NA
      } else{ # overlap with other terms
        terms.overlap[i] <- names(go.terms)[i]
      }
    }
    # exclude overlapped GO terms
    df.new <- df.GOenrich.res[!df.GOenrich.res$ID %in% terms.overlap,]
  } else if (nTerms == 1){
    df.new <- df.GOenrich.res
  } else if (nTerms <1){ # when 0 terms is significantly enriched
    df.new <- NA
  }
  return(df.new)
}

# GO enrichment analysis and plot non-overlapped top GO terms 
Calculate_topGO_plot <- function(data, num_topGO = 10, PATH.topGO, deg.sig, cols){
  clusters.TFoi <- unique(as.character(cds$meta.new))
  clusters.TFoi <- clusters.TFoi[clusters.TFoi != "Control"]
  erich.summary.clusters <- lapply(clusters.TFoi, function(x){
    erich.go.BP <- enrichGO(gene = deg.sig[deg.sig$cell_group == x, ]$gene_short_name,
                            OrgDb = org.Mm.eg.db,
                            keyType = "SYMBOL",
                            ont = "BP",
                            pvalueCutoff = 0.05,
                            qvalueCutoff = 0.05)
    erich.summary <- as.data.frame(erich.go.BP@result)
    erich.summary$cluster <- x
    erich.summary <- erich.summary %>% filter(p.adjust < 0.06) 
    cat("check the summary of GO_BP enrichment for ", x, ": \n")
    print(erich.summary)
    return(erich.summary)
  })
  
  # exclude overlapped GO withih each cluster
  df.non_overlap <- lapply(erich.summary.clusters, function(x) exclude_overlapped_GO(percent.overlap = 0.5,
                                                                                   df.GOenrich.res = x))
  df.non_overlap <- rbindlist(df.non_overlap)

  cat("!!! check if enriched GO terms are unique (FALSE) across clusters:", "\n")
  print(table(duplicated(df.non_overlap$ID)))
  

  df.non_overlap <- df.non_overlap %>% 
    group_by(cluster) %>% 
    arrange(p.adjust, .by_group = T) %>% 
    top_n(num_topGO, wt = -log10(p.adjust))
  
  data$meta.new <- fct_rev(data$meta.new)
  df.non_overlap$cluster <- factor(df.non_overlap$cluster, levels = levels(data$meta.new)[levels(data$meta.new) != "Control"])
  df.non_overlap <- df.non_overlap %>% 
    group_by(cluster) %>% 
    arrange(desc(p.adjust), .by_group = T) 
  
  
  Description.order <- df.non_overlap$Description
  df.non_overlap$Description <- factor(df.non_overlap$Description, levels = Description.order)
  

  p_GOenrich <- ggplot(df.non_overlap)+
    geom_bar(mapping = aes(x = -log10(p.adjust), y = Description, fill = cluster), 
             stat = "identity",  color = "white",
             width=0.5)+
    scale_fill_manual(values = cols)+
    theme_cowplot()+
    scale_x_continuous(position = "top")
  print(p_GOenrich)

  ggsave(p_GOenrich, file = PATH.topGO, width = 12, height = 6)
} 

# DE analysis and heatmap for Runx2 branching analysis
Runx2_DE_heatmap <- function(data.seu, data.gene_annot){
  DefaultAssay(data.seu) <- "RNA"
  
  # as no obvious batch effect was observed among Runx2 cells, used FindMarkers directly
  # high Runx2 vs low Runx2
  DEGs <- FindMarkers(data.seu, ident.1 = "high_Runx2",ident.2 = "low_Runx2", min.pct = 0.25, logfc.threshold = 0.25 )
  DEGs$pct.diff <- abs(DEGs$pct.1 - DEGs$pct.2)
  DEGs$gene <- data.gene_annot[rownames(DEGs), "gene_symbol"]
  DEGs <- DEGs[DEGs$p_val_adj < 0.05 & DEGs$pct.diff > 0.1,] # 0.1
  # high Runx2 vs control
  DEGs.high_VS_ctr <- FindMarkers(data.seu, ident.1 = "high_Runx2",ident.2 = "Control", min.pct = 0.25, logfc.threshold = 0.25 )
  DEGs.high_VS_ctr$pct.diff <- abs(DEGs.high_VS_ctr$pct.1 - DEGs.high_VS_ctr$pct.2)
  DEGs.high_VS_ctr$gene <- data.gene_annot[rownames(DEGs.high_VS_ctr), "gene_symbol"]
  DEGs.high_VS_ctr <- DEGs.high_VS_ctr[DEGs.high_VS_ctr$p_val_adj < 0.05 & DEGs.high_VS_ctr$pct.diff > 0.5,]
  # low Runx2 vs control
  DEGs.low_VS_ctr <- FindMarkers(data.seu, ident.1 = "low_Runx2",ident.2 = "Control", min.pct = 0.25, logfc.threshold = 0.25 )
  DEGs.low_VS_ctr$pct.diff <- abs(DEGs.low_VS_ctr$pct.1 - DEGs.low_VS_ctr$pct.2)
  DEGs.low_VS_ctr$gene <- data.gene_annot[rownames(DEGs.low_VS_ctr), "gene_symbol"]
  DEGs.low_VS_ctr <- DEGs.low_VS_ctr[DEGs.low_VS_ctr$p_val_adj < 0.05 & DEGs.low_VS_ctr$pct.diff > 0.25,]
  
  # take overlapped DE genes, which varied between high and low Runx2 cells but also between them and control
  genes.heatmap <- intersect(rownames(DEGs), unique(c(rownames(DEGs.low_VS_ctr), rownames(DEGs.high_VS_ctr))))
  
  # if check also endogeneous Runx2 expression (basal level)
  # genes.heatmap <- c(genes.heatmap, data.gene_annot[data.gene_annot$gene_symbol == "Runx2","ensembl_id"])
  cds <- GetAssayData(data.seu, assay = "RNA", slot = "data")
  heatmap_data <- cds[genes.heatmap,]
  rownames(heatmap_data) <- data.gene_annot[rownames(heatmap_data), "gene_symbol"]
  
  # order by Runx2 dose
  Order.byDose <- data.seu$Dose
  Order.byDose <- Order.byDose[order(Order.byDose, decreasing = F)] %>% names()
  heatmap_data <- heatmap_data[, Order.byDose]
  
  # prepare annotation and color scheme
  my_sample_col <- data.seu@meta.data[,"Runx2_dose", drop = F]
  my_sample_col$Dose <- data.seu$Dose
  
  my_colour = list(
    Runx2_dose = c(Control = "lightgrey", low_Runx2 = "#FF9933", high_Runx2 = "#996600"),
    Dose = rocket(200, direction = -1)
  )
  
  # plot gene expression as heatmap
  heatmap <- pheatmap::pheatmap(as.matrix(heatmap_data),
                                color = colorRampPalette(c("blue","white","red"))(200),
                                annotation_col = my_sample_col,
                                annotation_colors = my_colour,
                                border_color = NA, 
                                show_colnames = F, show_rownames = T,
                                cluster_cols = F, cluster_rows = T, 
                                scale = "row", 
                                clustering_method = "ward.D2",
                                annotation_legend = T
  )
  
  Output <- list(genes.heatmap, heatmap_data, heatmap)
  names(Output) <- c("genes.heatmap", "heatmap_data","heatmap_plot")
  return(Output)
}

```

## read the functional TF atlas in G1 (default phase)

```{r}
seu_funct <- readRDS("results/C3H10_10X_all_exps_integrated_functional_TF_atlas_G1_Phase.rds")
dim(seu_funct)
```

# Figure S4.1e - upper panel

```{r Fig_S4.1e_up}
TFois <- c("Egr1","Esr2","Etv1")
p_TFois <- lapply(TFois, function(x){
  p <- DimPlot(seu_funct, cells.highlight = colnames(seu_funct)[seu_funct$TF == x], cols = "gray85", cols.highlight = "red", sizes.highlight = 1, pt.size = 0.1)+
    theme(axis.line = element_blank(),
    legend.position = "none", 
    panel.grid.major=element_blank(), 
    panel.grid.minor=element_blank(), 
    axis.text=element_blank(), 
    axis.ticks=element_blank(), 
    plot.background=element_blank(), 
    panel.background=element_blank())+
    xlab(label = "") + ylab(label = "")+
    labs(title = x)
  return(p)
})
p_TFois

pdf("figures/Fig_S4.1e_up.pdf", width = 10, height = 8)
marrangeGrob(p_TFois, ncol = 2, nrow = 2)
dev.off()
```

# Figure S4.2d,e - upper panel & Figure S4.2g

```{r}
TFois <- c("Meis2","Myog","Grhl2")
p_TFois <- lapply(TFois, function(x){
  p <- DimPlot(seu_funct, cells.highlight = colnames(seu_funct)[seu_funct$TF == x],cols = "gray85", cols.highlight = "darkgreen", sizes.highlight = 1, pt.size = 0.1)+
    theme(axis.line = element_blank(),
    legend.position = "none", 
    panel.grid.major=element_blank(), 
    panel.grid.minor=element_blank(), 
    axis.text=element_blank(), 
    axis.ticks=element_blank(), 
    plot.background=element_blank(), 
    panel.background=element_blank())+
    xlab(label = "") + ylab(label = "")+
    labs(title = x)
  return(p)
})
p_TFois

pdf("figures/Fig_S4.2d_e_up_Fig_S4.2g.pdf", width = 10, height = 8)
marrangeGrob(p_TFois, ncol = 2, nrow = 2)
dev.off()
```

# cell fate branching analysis on (phase adjusted) G1 cells of TF of interest and batch-paired control

## read the annotation file

```{r}
data.annot <- fread("metadata/GRCm38_gene_annot.tsv", data.table = F) 
rownames(data.annot) <- data.annot$ensembl_id
```

## read the whole scTF-seq atlas

```{r}
seu <- readRDS("results/C3H10_10X_all_exps_merged_genefiltered_integrated_functional.rds")
dim(seu)
```

## subset to G1 (Phase adjusted) cells for TF of interest + batch-paired control

```{r}
# subset seurat object to G1
seu <- subset(seu, Phase_corrected == "G1")
DefaultAssay(seu) <- "RNA"
```

# Figure 4c-e and Figure S4.1d

```{r}
# re-analysis cells of interest
cds <- monocle3_reanalysis(tfoi = "Klf4", data = seu, data.gene_annot = data.annot, num_dims = 12, res = 0.2)
cds <- learn_graph(cds)
cds <- order_cells(cds) 

color.code <- c("lightgrey","#FF6600","blue")
cds$meta.new <- factor(cds$meta.new, levels = c("Control","3", "1"))
names(color.code) <- names(table(cds$meta.new))

# calculate DE
DEG.sig <- Calculate_DE(cds)
```

```{r Fig_4c_d}
# Fig_4c, d
plot_dose_clusters(data = cds, 
                   tfoi = "Klf4", 
                   PATH.UMAP = "figures/Fig_4c.pdf",
                   PATH.violin = "figures/Fig_4d.pdf",
                   cols = color.code)
# 	Welch Two Sample t-test
# 
# data:  meta.cluster$Dose[meta.cluster$meta.new == clusters.TFoi[1]] and meta.cluster$Dose[meta.cluster$meta.new == clusters.TFoi[2]]
# t = 4.5769, df = 63.851, p-value = 2.234e-05
# alternative hypothesis: true difference in means is not equal to 0
```

```{r Fig_4e}
# Fig_4e
Calculate_topGO_plot(data = cds,
                     num_topGO = 10,
                     PATH.topGO = "figures/Fig_4e.pdf",
                     deg.sig = DEG.sig, 
                     cols = color.code)
```

```{r Fig_S4.1d}
# Fig_S4.1d
genes_heatmap <- DEG.sig[order(DEG.sig$cell_group, decreasing = F),]$gene_id %>% unique()
heatmap_data <- exprs(cds)[genes_heatmap, ]
rownames(heatmap_data) <- data.annot[rownames(heatmap_data), "gene_symbol"]

cells.ctr <- colnames(cds)[cds$TF %in% c("D0","D0_confluent")]
Order.pseudotime_1 <- pseudotime(cds)[cells.ctr]
Order.pseudotime_1 <- Order.pseudotime_1[order(Order.pseudotime_1, decreasing = F)]

# order heatmap_data by pseudotime witin all TF cells
cells.oi <- colnames(cds)[cds$TF == "Klf4"]
Order.pseudotime_2 <- pseudotime(cds)[cells.oi]
Order.pseudotime_2 <- Order.pseudotime_2[order(Order.pseudotime_2, decreasing = F)]
Order.pseudotime <- c(Order.pseudotime_1,Order.pseudotime_2)
heatmap_data <- heatmap_data[, names(Order.pseudotime)] 

# prepare annotation for heatmap
my_sample_row <- cds@colData[,"Dose", drop = F] %>% as.data.frame()
my_sample_row$Pseudotime <- Order.pseudotime[rownames(my_sample_row)]
my_sample_row$Cluster <- colData(cds)$meta.new 

my_colour = list(
  Dose = rocket(200, direction = -1),
  Pseudotime = colorRampPalette(brewer.pal(9, "Blues"))(200),
  Cluster = c(Control = "lightgrey", '1' = "blue", '3' = "#FF6600") 
)

heatmap <- pheatmap::pheatmap(t(as.matrix(heatmap_data)),
                              color = colorRampPalette(c("blue","white","red"))(200),
                              annotation_row = my_sample_row,
                              annotation_colors = my_colour,
                              border_color = NA, 
                              show_colnames = T, show_rownames = F,
                              cluster_cols = F, cluster_rows = F, scale = "column", clustering_method = "ward.D2",
                              annotation_legend = T)
heatmap

ggsave(heatmap, file = "figures/Fig_S4.1d.pdf", width = 12, height = 3)

```


# Figure 4f-h

```{r}
# re-analysis cells of interest
cds <- monocle3_reanalysis(tfoi = "Grhl2", data = seu, data.gene_annot = data.annot, num_dims = 30, res = 0.3)
color.code <- c("lightgrey","green","#FF6600","blue")
cds$meta.new <- factor(cds$meta.new, levels = c("Control","5", "10", "1"))
names(color.code) <- names(table(cds$meta.new))
```

```{r Fig_4f_g}
plot_dose_clusters(data = cds, 
                   tfoi = TFoi, 
                   PATH.UMAP = "figures/Fig_4f.pdf",
                   PATH.violin = "figures/Fig_4g.pdf",
                   cols = color.code)
```

```{r Fig_4h}
DEG.sig <- Calculate_DE(cds)

Calculate_topGO_plot(data = cds,
                     num_topGO = 10,
                     PATH.topGO = "figures/Fig_4h.pdf",
                     deg.sig = DEG.sig, 
                     cols = color.code)
```

# Figure S4.1e (lower panel), f

```{r Fig_S4.1e-lower_f}
# Egr1
cds <- monocle3_reanalysis(tfoi = "Egr1", data = seu, data.gene_annot = data.annot, num_dims = 15, res = 0.1)
color.code <- c("lightgrey","#b3cde3", "orange")
cds$meta.new <- factor(cds$meta.new, levels = c("Control","1", "7"))
names(color.code) <- names(table(cds$meta.new))
p_Egr1 <- plot_dose_clusters(data = cds, 
                             tfoi = TFoi, 
                             PATH.UMAP = "results/na.pdf",
                             PATH.violin = "results/na.pdf",
                             cols = color.code,
                             output_plot = FALSE)
p_Egr1

# Esr2
cds <- monocle3_reanalysis(tfoi = "Esr2", data = seu, data.gene_annot = data.annot, num_dims = 15, res = 0.1)

color.code <- c("lightgrey","green", "#9999FF")
cds$meta.new <- factor(cds$meta.new, levels = c("Control","6", "4"))
names(color.code) <- names(table(cds$meta.new))
p_Esr2 <- plot_dose_clusters(data = cds, 
                             tfoi = TFoi, 
                             PATH.UMAP = "results/na.pdf",
                             PATH.violin = "results/na.pdf",
                             cols = color.code,
                             output_plot = FALSE)
p_Esr2

# Etv1
cds <- monocle3_reanalysis(tfoi = "Etv1", data = seu, data.gene_annot = data.annot, num_dims = 20, res = 0.1)
color.code <- c("lightgrey","#66cc99", "#CCCC00")
cds$meta.new <- factor(cds$meta.new, levels = c("Control","4", "6"))
names(color.code) <- names(table(cds$meta.new))
p_Etv1 <- plot_dose_clusters(data = cds, 
                             tfoi = TFoi, 
                             PATH.UMAP = "results/na.pdf",
                             PATH.violin = "results/na.pdf",
                             cols = color.code,
                             output_plot = FALSE)
p_Etv1

# Figure 4.1e-lower
ggsave(p_Egr1$umap.cluster+p_Esr2$umap.cluster+p_Etv1$umap.cluster, 
       file = "figures/Fig_S4.1e-lower.pdf", width = 14, height = 4)


# Figure 4.1e-lower
ggsave(p_Egr1$dose.violin+p_Esr2$dose.violin+p_Etv1$dose.violin, 
       file = "figures/Fig_S4.1f.pdf", width = 12, height = 4)

```

# Figure S4.2d, e (lower panel), f

```{r Fig_S4.2d_e-lower_f}
# Meis2
cds <- monocle3_reanalysis(tfoi = "Meis2", data = seu, data.gene_annot = data.annot, num_dims = 30, res = 0.2)
color.code <- c("lightgrey","#7570b3", "#cab2d6","#99FF99","#CCCC00")
cds$meta.new <- factor(cds$meta.new, levels = c("Control","5", "9", "8", "3"))
names(color.code) <- names(table(cds$meta.new))
p_Meis2 <- plot_dose_clusters(data = cds, 
                              tfoi = "Meis2", 
                              PATH.UMAP = "results/na.pdf",
                              PATH.violin = "results/na.pdf",
                              cols = color.code,
                              output_plot = FALSE)
p_Meis2
# 	Pairwise comparisons using t tests with pooled SD 
# 
# data:  meta.cluster$Dose and meta.cluster$meta.new 
# 
#   5       9       8      
# 9 0.00068 -       -      
# 8 1.0e-05 0.44412 -      
# 3 2.7e-08 0.09367 0.27909
# 
# P value adjustment method: fdr 


# Myog
cds <- monocle3_reanalysis(tfoi = "Myog", data = seu, data.gene_annot = data.annot, num_dims = 12, res = 0.01)
color.code <- c("lightgrey","#cab2d6", "#984ea3","#cc99ff")
cds$meta.new <- factor(cds$meta.new, levels = c("Control","4", "6", "3"))
names(color.code) <- names(table(cds$meta.new))
p_Myog <- plot_dose_clusters(data = cds, 
                              tfoi = "Myog", 
                              PATH.UMAP = "results/na.pdf",
                              PATH.violin = "results/na.pdf",
                              cols = color.code,
                              output_plot = FALSE)
p_Myog
# 	Pairwise comparisons using t tests with pooled SD 
# 
# data:  meta.cluster$Dose and meta.cluster$meta.new 
# 
#   4       6   
# 6 1.9e-06 -   
# 3 1.5e-05 0.46
# 
# P value adjustment method: fdr 


# Figure 4.1e-lower
ggsave(p_Meis2$umap.cluster+p_Myog$umap.cluster, 
       file = "figures/Fig_S4.2d_e-lower.pdf", width = 10, height = 4)


# Figure 4.1e-lower
ggsave(p_Meis2$dose.violin+p_Myog$dose.violin,
       file = "figures/Fig_S4.2f.pdf", width = 10, height = 4)

```

# Figure S4.2h

```{r Fig_S4.2h}
cds <- monocle3_reanalysis(tfoi = "Grhl2", data = seu, data.gene_annot = data.annot, num_dims = 30, res = 0.3)
# learn pseudotime using cells in control clusters as roots
cds <- learn_graph(cds)
cds <- order_cells(cds) 

p.pseudo <- plot_cells(cds, color_cells_by = "pseudotime", cell_size = 1, show_trajectory_graph = F)+
  geom_point(aes(color = cell_color), size = 1, stroke = 1.5)+
  theme_void()
p.pseudo

ggsave(p.pseudo, filename = "figures/Fig_S4.2g.pdf", width = 5, height = 3.5)
```

# Figure S4.2a-c branching analysis for Runx2

```{r Fig_S4.2a-c}
TFoi <- "Runx2"

# subset to Runx2 cells + batch-paired control
batch.oi <- table(seu$batch[seu$TF == TFoi]) %>% names()
seu.oi <- subset(seu, TF %in% c("D0", "D0_confluent", TFoi) & batch %in% batch.oi)

# group Runx2 cells based on dose cutoff
summary(seu.oi$Dose[seu.oi$TF == TFoi])
 #   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
 # 0.6931  1.6094  2.3979  2.4736  3.2581  6.1821 
seu.oi$Runx2_dose <- seu.oi$TF
seu.oi$Runx2_dose[seu.oi$TF != "Runx2"] <- "Control"
seu.oi$Runx2_dose[seu.oi$TF == "Runx2" & seu.oi$Dose >= 2.5] <- "high_Runx2"
seu.oi$Runx2_dose[seu.oi$TF == "Runx2" & seu.oi$Dose < 2.5] <- "low_Runx2"
seu.oi$Runx2_dose <- factor(seu.oi$Runx2_dose, levels = c("Control","low_Runx2","high_Runx2"))
Idents(seu.oi) <- "Runx2_dose"


# re-run UMAP on integrated assay
DefaultAssay(seu.oi) <- "integrated"
set.seed(10)
seu.oi <- seu.oi %>% RunPCA(npcs = 50) %>% RunUMAP(dims = 1:50)


# Figure S4.2a
p_dose <- FeaturePlot(seu.oi, features = "Dose", order = T)+scale_color_viridis()+
    theme(axis.line = element_blank(), 
    panel.grid.major=element_blank(), 
    panel.grid.minor=element_blank(), 
    axis.text=element_blank(), 
    axis.ticks=element_blank(), 
    plot.background=element_blank(), 
    panel.background=element_blank())+
    xlab(label = "") + ylab(label = "")+
  labs(title = paste0(TFoi," dose"))
p_dose
ggsave(p_dose, filename = "figures/Fig_S4.2a.pdf", width = 6, height = 5)

# Figure S4.2b
p_group <- DimPlot(seu.oi, group.by = "Runx2_dose", cols = c("lightgray", "#FF9933","#996600"))+
    theme(axis.line = element_blank(), 
    panel.grid.major=element_blank(), 
    panel.grid.minor=element_blank(), 
    axis.text=element_blank(), 
    axis.ticks=element_blank(), 
    plot.background=element_blank(), 
    panel.background=element_blank())+
    xlab(label = "") + ylab(label = "")+
  labs(title = TFoi)
p_group
ggsave(p_group, filename = "figures/Fig_S4.2b.pdf", width = 7, height = 5)

# Figure S4.2c
final_output <- Runx2_DE_heatmap(data.seu = seu.oi, data.gene_annot = data.annot) 
final_output$heatmap_plot
ggsave(final_output$heatmap_plot, filename = "figures/Fig_S4.2c.pdf", width = 6.5, height = 5)

```
